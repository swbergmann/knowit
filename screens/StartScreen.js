import { useState, useEffect } from 'react';
import { StyleSheet, Text, TextInput, View, Image, Pressable, Keyboard, KeyboardAvoidingView, SafeAreaView, TouchableWithoutFeedback, Platform } from 'react-native';
import { SimpleLineIcons } from "@expo/vector-icons";

import Colors from '../constants/colors';

// use object-destructuring to automatically pull out these props from the incoming props object
function StartScreen({onStore, onGetName}) {
  const [userInput, setUserInput] = useState('');
  const [errors, setErrors] = useState({});
  const [isFormValid, setIsFormValid] = useState(false);

  function inputChangeHandler(text) {
    setUserInput(text.trim()); // remove white-spaces
  }

  function pressHandler() {
    if (isFormValid) { // only store input if form is validated
      onStore(userInput);
      onGetName(); // update storedName to switch screens
    }
  }

  const validateForm = () => {
    let errors = {};

    if (!userInput) { // validate name field
      errors.name = 'Name is required.';
    } else if (userInput.length < 3) {
      errors.name = 'Name must be at least 3 characters.';
    } else if (userInput.length > 12) {
      errors.name = 'Maximum length is 12 characters.';
    }

    setErrors(errors);
    setIsFormValid(Object.keys(errors).length === 0); // form is only valid if 0 errors occur
  }

  useEffect(() => {
    validateForm(); // validate form on every user input
  }, [userInput]);

  return (
    <KeyboardAvoidingView
      behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
      style={styles.container}>
      <SafeAreaView style={styles.safe}>
        <TouchableWithoutFeedback onPress={Keyboard.dismiss}>
          <View style={styles.inner}>
            <View style={styles.visible}>
              <View style={styles.upper}>
                <Image 
                  source={require('../assets/undraw/welcome_cats.png')}
                  style={styles.welcomeImage}
                  resizeMode='contain'
                />

                <Text style={styles.knowit}>Know It?!</Text>
              </View>
              <View style={styles.lower}>
                <Text style={styles.description}>Enter your name</Text>

                <View style={styles.inputLine}>
                  <SimpleLineIcons name='graduation' color='black' style={styles.icon}/>
                  <TextInput 
                    style={styles.input}
                    placeholder='Name'
                    onChangeText={inputChangeHandler}
                    />
                </View>

                {/* show errors only when they exist */}
                {Object.values(errors).map((error, index) => (
                  <Text key={index} style={styles.error}>
                    {error}
                  </Text>
                ))}

                <Pressable 
                  style={styles.button}
                  onPress={pressHandler}
                >
                  <Text style={styles.white}>Login</Text>
                </Pressable>
              </View>
            </View>
            <View style={styles.empty}></View>
          </View>
        </TouchableWithoutFeedback>
      </SafeAreaView>
    </KeyboardAvoidingView>
  );
}

export default StartScreen;

const styles = StyleSheet.create({
    container: {
      flex: 1,
      backgroundColor: Colors.primary600
    },
    safe: {
      flex: 1,
      marginTop: 90
    },
    inner: {
      flex: 1,
      margin: 20,
      overflow: 'hidden',
      justifyContent: 'flex-end',
      borderRadius: 8
    },
    visible: {
      overflow: 'hidden',
      borderRadius: 8
    },
    upper: {
      alignItems: 'center',
      width: '100%',
      backgroundColor: 'white',
      borderBottomWidth: 2,
      borderBottomColor: Colors.gray400
    },
    lower: {
      alignItems: 'center',
      backgroundColor: Colors.gray000,
      paddingVertical: 20
    },
    empty: {
      flex: 1
    },
    welcomeImage: {
      width: '100%',
      height: 210
    },
    knowit: {
      fontSize: 36,
      marginVertical: 20
    },
    description: {
      fontSize: 20,
      color: Colors.gray600
    },
    inputLine: {
      flexDirection: 'row',
      borderWidth: 2,
      borderColor: Colors.gray400,
      backgroundColor: 'white',
      borderRadius: 6,
      alignSelf: 'stretch',
      marginHorizontal: 24,
      marginVertical: 24,
      padding: 10,
      overflow: 'hidden'
    },
    icon: {
      marginTop: 2,
      fontSize: 24
    },
    input: {
      fontSize: 24,
      marginHorizontal:24,
      width: '100%'
    },
    button: {
      backgroundColor: Colors.primary200,
      alignItems: 'center',
      justifyContent: 'center',
      paddingVertical: 12,
      paddingHorizontal: 32,
      marginVertical: 32,
      marginTop: 85,
      borderRadius: 6
    },
    white: {
      color: 'white',
      fontSize: 24
    },
    error: {
      color: 'red',
      fontSize: 20
    }
});