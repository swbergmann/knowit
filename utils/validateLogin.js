/**
 * This function is located in the '/utils' so that it can be tested by the unit tests
 * 
 * validateLogin
 * - takes 'input' (which is the user input in a text field) as a parameter.
 * - returns an empty string in case the input value complies with all validations.
 * - returns an error message in case the input value must be corrected by the user
 */
export const validateLogin = (input) => {
    let error = '';
    let testInput = input.replaceAll(' ', '');  // remove all empty spaces from input

    if (!testInput) {                           // input is empty
      error = 'Name is required.';
    } else if (testInput.length < 3) {          // input is too short
      error = 'Minimum length is 3 characters.';
    } else if (testInput.length > 10) {         // input is too long
      error = 'Maximum length is 10 characters.';
    }

    return error;
}