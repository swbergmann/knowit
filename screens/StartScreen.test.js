import React from 'react';
import renderer from 'react-test-renderer';

import StartScreen from "./StartScreen";
import { validateLogin } from '../utils/validateLogin';

/**
 * To execute unit tests in the terminal run:
 * npm test
 * 
 * In case the UI component changes, update the component snapshot:
 * npm run test -- -u
 */
it('renders StartScreen correctly', () => {
  const tree = renderer
    .create(<StartScreen />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

/**
 * The validateLogin function is tested with unit tests to ensure application users
 * get helpful interactive feedback and can correct their faulty input values easily.
 * 
 * validateLogin is called with several edge-case string values as a parameter and
 * the expected return value of validateLogin is compared with the assumed return value.
 * 
 * It is critical for the user interface, that the user name is not too short and not too long.
 */
it('does not allow empty input for login', () => {
  expect(validateLogin(''))
  .toBe('Name is required.');
});

it('does not allow blank spaces as input for login', () => {
  expect(validateLogin('     '))
  .toBe('Name is required.');
});

it('does not allow 1 letter input for login', () => {
  expect(validateLogin('S'))
  .toBe('Name must be at least 3 characters.');
});

it('does not allow 2 letter input for login', () => {
  expect(validateLogin('Se'))
  .toBe('Name must be at least 3 characters.');
});

it('does not allow several blank spaces in combination with 2 letter input for login', () => {
  expect(validateLogin(' S e '))
  .toBe('Name must be at least 3 characters.');
});

it('allows 3 letter input for login', () => {
  expect(validateLogin('Seb'))
  .toBe('');
});

it('allows 9 letter input for login', () => {
  expect(validateLogin('Sebastian'))
  .toBe('');
});

it('allows 10 letter input for login', () => {
  expect(validateLogin('SebastianS'))
  .toBe('');
});

it('does not allow more than 10 letter input for login', () => {
  expect(validateLogin('SebastianSe'))
  .toBe('Maximum length is 10 characters.');
});

it('does not allow more than 10 letter input for login', () => {
  expect(validateLogin('SebastianSebastianSebastian'))
  .toBe('Maximum length is 10 characters.');
});