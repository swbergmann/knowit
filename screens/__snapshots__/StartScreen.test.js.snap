// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders StartScreen correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    [
      {
        "backgroundColor": "#18254D",
        "flex": 1,
      },
      {
        "paddingBottom": 0,
      },
    ]
  }
>
  <RCTSafeAreaView
    style={
      {
        "flex": 1,
        "marginTop": 60,
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "borderRadius": 8,
          "flex": 1,
          "justifyContent": "flex-end",
          "margin": 20,
          "overflow": "hidden",
        }
      }
    >
      <View
        style={
          {
            "borderRadius": 8,
            "overflow": "hidden",
          }
        }
      >
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "white",
              "borderBottomColor": "#d5d5d5",
              "borderBottomWidth": 2,
              "width": "100%",
            }
          }
        >
          <Image
            resizeMode="contain"
            source={1}
            style={
              {
                "height": 210,
                "width": "100%",
              }
            }
          />
          <Text
            style={
              {
                "fontSize": 24,
              }
            }
          >
            to the quizzing app
          </Text>
          <Text
            style={
              {
                "fontSize": 36,
                "marginVertical": 20,
              }
            }
          >
            Know It?!
          </Text>
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "#f1f3f5",
              "paddingVertical": 20,
            }
          }
        >
          <Text
            style={
              {
                "color": "#878787",
                "fontSize": 20,
              }
            }
          >
            Enter your name
          </Text>
          <View
            style={
              {
                "alignSelf": "stretch",
                "backgroundColor": "white",
                "borderColor": "#d5d5d5",
                "borderRadius": 6,
                "borderWidth": 2,
                "flexDirection": "row",
                "marginHorizontal": 24,
                "marginVertical": 24,
                "overflow": "hidden",
                "padding": 10,
              }
            }
          >
            <Text />
            <TextInput
              autoCorrect={false}
              onChangeText={[Function]}
              placeholder="Name"
              style={
                {
                  "fontSize": 24,
                  "marginHorizontal": 24,
                  "width": "100%",
                }
              }
            />
          </View>
          <Text
            style={
              {
                "color": "red",
                "fontSize": 20,
              }
            }
          />
          <View
            style={
              {
                "borderRadius": 6,
                "elevation": 2,
                "marginTop": 85,
                "marginVertical": 32,
                "overflow": "visible",
                "shadowColor": "black",
                "shadowOffset": {
                  "height": 2,
                  "width": 0,
                },
                "shadowOpacity": 0.15,
                "shadowRadius": 8,
              }
            }
          >
            <View
              style={
                {
                  "borderRadius": 6,
                  "overflow": "hidden",
                }
              }
            >
              <View
                accessibilityState={
                  {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": undefined,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#4B58DD",
                    "justifyContent": "center",
                    "paddingHorizontal": 12,
                    "paddingVertical": 8,
                    "width": 120,
                  }
                }
              >
                <Text
                  style={
                    {
                      "color": "white",
                      "fontSize": 24,
                    }
                  }
                >
                  Login
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          {
            "flex": 1,
          }
        }
      />
    </View>
  </RCTSafeAreaView>
</View>
`;
